### DISCUSSION TITLE
neovim Copilot inserts weird character sometimes

### DISCUSSION THREAD

**Discussion by Jacfger** [community/community#29817] [Copilot Conversations] [answered]
I'm now currently using Copilot.vim 1.4.3, and I've tried 1.5.0, 1.5.1 as well. I haven't tried other versions but all of them had the problem where sometimes when I confirm the strings, ```<80>kD``` is inserted at the beginning of the strings. 

After some Googling I found that it's probably a backspace character, but my keybinding doesn't contains a backspace, (it's alt+N, or Meta-N. I've also tried <right> so it shouldn't be a problem). 

Do anyone has any idea on how to fix this?

**Comment by fragov** [comment]
I have the same issue, sometimes copilot add '<80>kl<80>kD<80>kl<80' before result.

**Comment by Leixb** [comment]
I had the same issue when using copilot alongside [nvim-cmp](https://github.com/hrsh7th/nvim-cmp). Setting the `copilot#Accept` throught `nvim-cmp` as explained in the [`nvim-cmp` help](https://github.com/hrsh7th/nvim-cmp/blob/b16e5bcf1d8fd466c289eab2472d064bcd7bab5d/doc/cmp.txt#L830-L852) seems to have solved it. 

Here is the relevant portion of my configuration:

```lua
vim.g.copilot_no_tab_map = true
vim.keymap.set(
    "i",
    "<Plug>(vimrc:copilot-dummy-map)",
    'copilot#Accept("")',
    { silent = true, expr = true, desc = "Copilot dummy accept" }
)

local cmp = require("cmp")
cmp.setup {
  mapping = {
    ['<M-;>'] = cmp.mapping(function(fallback)
      vim.api.nvim_feedkeys(vim.fn['copilot#Accept'](vim.api.nvim_replace_termcodes('<Tab>', true, true, true)), 'n', true)
    end)
  },
}

```

**Reply by pidgeon777** [reply]
This solved the problem. Thank you. 👍

**Comment by ariel-frischer** [comment]
@Leixb Don't you need to add copilot as a nvim-cmp source? This doesn't seem to work for me :/

**Reply by Leixb** [reply]
No, I don't have it as a source. You can check my full cmp and copilot configuration here: 
- https://github.com/Leixb/nixos-dotfiles/blob/master/users/leix/neovim/cmp.lua
- https://github.com/Leixb/nixos-dotfiles/blob/master/users/leix/neovim/copilot-vim.lua

**Comment by ArtursTimofejevs** [comment]
If anyone else is stumbling on this issue then this is how I solved it:

- I had used `vim.keymap.set` in a local variable as a reusable function called map and this seems to cause issues
```lua
local map = vim.keymap.set
map("i", "<C-j>", "copilot#Accept('<CR>')", {noremap = true, silent = true, expr=true })
```
- I changed to using `vim.api.nvim_set_keymap`
```lua
local map = function (type, key, value, opts)
  local options = { noremap = true }
  if opts then options = vim.tbl_extend('force', options, opts) end
  vim.api.nvim_set_keymap(type, key, value, options)
end
```

Seems to have fixed the issue :)

**Reply by nklsla** [reply]
Perfect, thanks!

**Comment by torarvid** [comment, marked as answer]
I tried understanding why @ArtursTimofejevs's suggestion worked, and I **think** I might have found something in the help for `vim.keymap.set` when looking at the `{opts}` (the `{noremap = true, silent = true, expr=true }` in the example)

> Accepts options accepted by the {opts} parameter in
                  |nvim_set_keymap()|, with the following notable differences:
                  • replace_keycodes: Defaults to `true` if "expr" is `true`.

So I was able to fix it by keeping the use of `vim.keymap.set` but just adding `, replace_keycodes = false`:

```lua
local map = vim.keymap.set
map("i", "<C-j>", "copilot#Accept('<CR>')", {noremap = true, silent = true, expr=true, replace_keycodes = false })
```

I believe this fixed the issue for me.

**Reply by dsuttonpreece** [reply]
I have a different setup, but adding `replace_keycodes = false` seems to have worked for me as well :)

**Reply by wmeints** [reply]
I tried this as well and it works perfectly. Thanks for the tip!

**Reply by devstefancho** [reply]
You save my time! Big Thanks!!

**Reply by levifig** [reply]
This is the way! ✊

**Reply by isvicy** [reply]
Thanks a lot! I use NvChad as my base config. This works perfectly!

**Reply by nburnett** [reply]
This solved my issue using Astrovim as my base config, Thank you!

**Reply by efficacy38** [reply]
this also works for me, thanks a lots!

**Reply by phrmendes** [reply]
amazing! thanks a lot ❤️❤️❤️❤️

**Reply by paulo-granthon** [reply]
finally after hours, managed to configure copilot.vim with C-a as accept thanks to your comment. THANKS!

**Reply by gmtborges** [reply]
Thank you, man!! Save my day

**Reply by farnots** [reply]
Working great thanks a lot
